services/iplant-groups/src/iplant_groups/clients/grouper.clj

```diff
@@ -611,25 +642,80 @@
         first)))
 
 (defn assign-role-permission
-  [username attribute-def-name-id role-id allowed? action-names]
+  [username attribute-def-name role-name allowed? action-names]
   (assign-remove-permission
     (format-role-permission-assign-request
-      username attribute-def-name-id role-id allowed? action-names)))
+      username attribute-def-name role-name allowed? action-names)))
 
 (defn remove-role-permission
-  [username attribute-def-name-id role-id action-names]
+  [username attribute-def-name role-name action-names]
   (assign-remove-permission
     (format-role-permission-remove-request
-      username attribute-def-name-id role-id action-names)))
+      username attribute-def-name role-name action-names)))
 
 (defn assign-membership-permission
-  [username attribute-def-name-id role-id subject-id allowed? action-names]
+  [username attribute-def-name role-name subject-id allowed? action-names]
   (assign-remove-permission
     (format-membership-permission-assign-request
-      username attribute-def-name-id role-id subject-id allowed? action-names)))
+      username attribute-def-name role-name subject-id allowed? action-names)))
 
 (defn remove-membership-permission
-  [username attribute-def-name-id role-id subject-id action-names]
+  [username attribute-def-name role-name subject-id action-names]
   (assign-remove-permission
     (format-membership-permission-remove-request
-      username attribute-def-name-id role-id subject-id action-names)))
+      username attribute-def-name role-name subject-id action-names)))
+
+(defn- format-all-permission-search-request
+  [username attribute-def-name]
+  {:WsRestGetPermissionAssignmentsRequest
+   (remove-vals nil? {:actAsSubjectLookup (act-as-subject-lookup username)
+                      :wsAttributeDefNameLookups [{:name attribute-def-name}]})})
+
+(defn- get-permission-assign-ids
+  [username attribute-def-name]
+  (->> (format-all-permission-search-request username attribute-def-name)
+       (permission-assignment-search*)
+       (distinct-by :attributeAssignId)
+       (group-by :permissionType)
+       (map-kv (fn [k v] [k (mapv :attributeAssignId v)]))))
```

Comment:
Each attribute assignment is listed multiple times because of the way that Grouper expands permissions. This function returns a list of distinct permission assignment identifiers grouped by permission type. This allows us to at least somewhat reliably remove all permissions associated with a resource using the smallest number of Grouper web service calls possible (one service call for the lookup, one service call to remove all role permissions (if there are any) and one service call to remove all role-subject permissions (if there are any).